### PR TITLE
Add PING monitor for backend liveness checks

### PR DESCRIPTION
**Pull request by Chaho12**
## Description

Add PING monitor for backend liveness checks

## Additional context and related issues

https://github.com/trinodb/trino-gateway/issues/1171

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
(x) Release notes are required, with the following suggested text:

```markdown
* Add PING monitor for backend liveness checks
```

### CODE REVIEW

**Review by ebyhr — CHANGES_REQUESTED** [review]
Please add integration tests with Kyuubi.

```
GET /v1/info	❌	INFO_API monitor	not implemented
```

Why not contribute to Kyuubi?

**Comment by Chaho12** [comment]
> Please add integration tests with Kyuubi.

Isn't it too much for simple ping-pong? I mean Kyuubi is an example that needs ping-pong, but I thought it is too early stage to do integration test with Kyuubi.

> Why not contribute to Kyuubi?

Since healthcheck is meant for client(Trino Gateway), and Kyuubi just supports simple health Ping-Pong health check, it is upto Gateway to check if backend cluster(kyuubi) is alive. So, I thought Gateway should have simple health check instead.

**Comment by RoeyoOgen** [comment]
@coderabbitai

**Comment by RoeyoOgen** [comment]
> Tip
> 
> For best results, initiate chat on the files or code changes.
> 
> 🧩 Analysis chain
> 🏁 Script executed:
> 
> ```shell
> #!/bin/bash
> gh pr view 1172 --repo trinodb/trino-gateway --json title,body,comments,reviews --jq '.'
> ```
> 
> Length of output: 3073
> 
> `@RoeyoOgen`, I don't have a strong stake in this repo's testing conventions, but here's a pragmatic take on the back-and-forth between Chaho12 and ebyhr:
> 
> **Both points have merit:**
> 
> * ebyhr's ask for integration tests is reasonable in general — testing a liveness monitor against a real (or realistic) backend gives higher confidence than unit tests with mocks alone, especially for network/protocol edge cases (timeouts, malformed responses, connection resets).
> * Chaho12's pushback is also valid — spinning up a full Kyuubi integration test harness is a heavier lift than the feature warrants at this stage, and the health-check contract here is intentionally generic (any backend exposing a simple ping endpoint), not Kyuubi-specific.
> 
> **A middle-ground suggestion:** Rather than a full Kyuubi integration test, consider adding a lightweight test that:
> 
> * Spins up a minimal in-process HTTP/TCP server (e.g., using a test utility server or `MockWebServer`-style helper) that mimics a ping endpoint's success/failure/timeout responses.
> * Verifies the `PingMonitor` correctly interprets healthy vs. unhealthy vs. unreachable states against that fake server.
> 
> This gets you integration-level confidence in the monitor's actual network behavior without requiring a real Kyuubi dependency in CI, and can be extended later if/when a proper Kyuubi integration suite is added. If the maintainers still want a true Kyuubi-backed test eventually, that could be tracked as a follow-up rather than blocking this PR.
> 
> Would it help if I opened a follow-up issue to track "Add Kyuubi-based integration test for PING monitor" so this PR isn't blocked on it? Let me know and I can file it with the appropriate context and backlink.

@Chaho12 what do you think about his suggestions?

@coderabbitai pls review the code as well

**Comment by Chaho12** [comment]
> what do you think about his suggestions?

There is the integration test for spinning a HttpServer serving ping pong test at `TestClusterStatsPingMonitor.java`.
I added some more test on it.

Regarding integration test with real Kyuubi container, what about a separate PR, that introduces a kyuubi container image + service config.

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsPingMonitor.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+            Request request = prepareGet()
+                    .setUri(uriBuilderFrom(URI.create(baseUrl)).appendPath(pingPath).build())
+                    .build();
+            StatusResponse response = client.execute(request, createStatusResponseHandler());
+            if (response.getStatusCode() == OK.code()) {
+                return TrinoStatus.HEALTHY;
+            }
+            if (shouldRetry(response.getStatusCode()) && retriesRemaining > 0) {
+                log.warn("Retrying ping health check on status %d for %s", response.getStatusCode(), request.getUri());
+                return checkStatus(baseUrl, retriesRemaining - 1);
```
Comment (RoeyoOgen) [inline gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsPingMonitor.java:83]: just a thought: would we prefer a simple loop instead of recursion?

File: gateway-ha/src/test/java/io/trino/gateway/ha/clustermonitor/TestClusterStatsMonitor.java [hunk 2]
```diff
@@ -162,6 +162,54 @@ void testMetricsMonitor()
                 new MonitorConfiguration()));
     }
 
+    @Test
```
Comment (RoeyoOgen) [inline gateway-ha/src/test/java/io/trino/gateway/ha/clustermonitor/TestClusterStatsMonitor.java:165]: i might be missing something but Current tests never verify that retry logic actually works.

perhpse add a test:
First response: 503
Second response: 200
Assert final status is HEALTHY

**Comment by andythsu** [comment]
I'm not familiar with Kyuubi but it looks like it's a gateway just like Trino Gateway

The intention is that we want to put Trino Gateway in front of Kyuubi, which is in front of trino clusters

So it flows like Trino Gateway -> Kyuubi -> Trino cluster?

If I understand correctly, is it meaningful to return a simple 200? Because if Kyuubi is simply a gateway, the Trino clusters behind it it could be dead while Kyuubi is still responding with 200, making this health check virtually useless.

Am I understanding this architecture correctly? Would it be more meaningful to add a health check that actually probes the actual Trino cluster? 

At the end of the day, Trino Gateway was intended to be placed in front of Trino clusters, and the health check is the source of truth that reports the health status of Trino clusters. It's fine to chain Trino Gateway to another gateway, but if it doesn't get the source of truth, it's practically not useful.

**Comment by Chaho12** [comment]
> So it flows like Trino Gateway -> Kyuubi -> Trino cluster?

@andythsu 
I explained it in more detailed in the #1171 , but the flow is as following
So, it practically integrates one (trino) cli to query both Trino SQL and Spark SQL. 
This would then allow setting routing rules for Kyuubi/Spark too!

For example, you can set so that if you do INSERT query (which is very heavy load in Trino) so why not run it in Spark automatically? And If you do SELECT, why wait for Spark to load, but rather let it run in Trino which is much faster? User only wants the results and is not that interested in the how part.

There are much steps to go to handle but this is just a start.

## AS-IS
```
trino CLI  ->  Kyuubi  -> trino
trino CLI -> trino gateway -> trino
```
## TO-BE

```
trino CLI  ->  trino-gateway  ->  Kyuubi  ->  Spark on YARN
                              \->  Trino (or another Kyuubi)
```
OR
```
trino CLI  ->  Kyuubi ->  Spark on YARN
                                  \-> trino-gateway-> Trino (or another Kyuubi)
                                  \-> trino
```

**Comment by Chaho12** [comment]
### FYI: how Kyuubi picks an engine

Kyuubi does not inspect the SQL. The engine comes from one session conf,
`kyuubi.engine.type`, resolved as:

1. Conf the client sends at connect time (JDBC URL)
2. Per-user default in `kyuubi-defaults.conf` (`___<user>___.` prefix)
3. Server default — `SPARK_SQL`

...then overridden by the `SessionConfAdvisor` plugin, which has the
final say. Valid values: `SPARK_SQL`, `FLINK_SQL`, `TRINO`, `HIVE_SQL`,
`JDBC`, `DATA_AGENT`. All share the same Thrift port; only the conf
differs:
```
    beeline -u 'jdbc:hive2://kyuubi:10009/;#kyuubi.engine.type=TRINO'
    beeline -u 'jdbc:hive2://kyuubi:10009/;#kyuubi.engine.type=HIVE_SQL'
```

### Why it matters here

- **`engine.type` is per session, not per statement.** Kyuubi will not
  split one session's statements across engines, so "INSERT on Spark,
  SELECT on Trino" has to be decided *before* the session opens — by the
  gateway or the client, not by Kyuubi.
- **But it is overridable server-side.** A `SessionConfAdvisor` plugin
  can rewrite `kyuubi.engine.type` per user/session and beats whatever
  the client sent. That is the hook if we want engine policy centralized
  in Kyuubi rather than in the gateway.
- **A 200 from Kyuubi does not mean the engine works.** Engine identity
  is `user + engineType + shareLevel + subdomain`, and the engine is only
  launched when a session of that type opens. A Kyuubi frontend can pass
  `PING` while its Trino engine cannot start (missing
  `kyuubi.session.engine.trino.connection.url` / `.connection.catalog`)
  or its Hive engine has no `HIVE_HOME`. Re: @ebyhr's question — `PING`
  proves the frontend is up, not that any engine behind it is usable.
